### ISSUE TITLE
Reusing nodes should not retain the class list

### ISSUE THREAD

**Issue by bionicbrian** [developit/preact#656]
(Originally created in preact-compat as https://github.com/developit/preact-compat/issues/365. See that issue for link to possible culprit.)

In 8.1.0 with `preact-compat@3.15.0`, it appears that render reuses the first child of the target node _and_ its class list? But _only_ when the first element of the react component tree does _not_ have a class list?

Simply applying `className=""` to the first element of the top-level React component solves the error, but should probably not be necessary?

This can cause some weird styling due to unexpected CSS inheritance. 

Example:
https://codepen.io/bionicbrian/pen/GmpzVP

I think this may be related to this preact issue: https://github.com/developit/preact/issues/544 ?

**Comment by NekR** [comment]
@developit I think that's the same as I have with changing a property on a DOM node directly and then if that property isn't controlled by JSX/`h()` attributes -- it persist for next `preact.render()` mounts.

I know it's component recycling. :-(

**Comment by developit** [comment]
In his case it's picking up from SSR, since preact [no longer does a diff](https://github.com/developit/preact/commit/25f462b315f307a56f7a59265eeffadcd8569722#diff-a0896fa5d65b1d80ca4d496badbae50fL130) and instead just blasts properties out to the DOM.

**Comment by timmywil** [comment]
I think I'm having this issue as well. I have a structure that is basically:

```html
<div className='container'>
  {somecomponent}{/* resolves to <div></div>*/}
  <div className='error-banner'></div>
</div>
```

That first child div from "somecomponent" is getting the `error-banner` class, which ends up hiding it. This only happens when I swap react for preact-compat.

**Comment by timmywil** [comment]
I can confirm that adding a class, like below, fixes the issue.

```html
<div className='container'>
  {somecomponent}{/* resolves to <div className='something'></div>*/}
  <div className='error-banner'></div>
</div>
```

**Comment by developit** [comment]
yup. right now preact does not remove DOM attributes it does't have replacement values for when performing an initial render. If anyone wants to PR adding the diff back in that'd be awesome, otherwise I'll try to get to that as soon as I can.

**Comment by developit** [comment]
Fixed in #677.
